File: Java/E-montazysta/src/main/java/com/emontazysta/model/Element.java
package com.emontazysta.model;

import com.emontazysta.enums.TypeOfUnit;
import com.fasterxml.jackson.annotation.JsonIdentityInfo;
import com.fasterxml.jackson.annotation.ObjectIdGenerators;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.SQLDelete;

import javax.persistence.*;
import java.util.List;
import java.util.Optional;

@Data
@Entity
@Builder
@NoArgsConstructor
@AllArgsConstructor
@SQLDelete(sql = "UPDATE element SET deleted = true WHERE id=?")
@JsonIdentityInfo(generator = ObjectIdGenerators.PropertyGenerator.class, property = "id")
public class Element {

    public Element(Long id, String name, String code, TypeOfUnit typeOfUnit, float quantityInUnit,
                   List<ElementReturnRelease> elementReturnReleases, List<ElementInWarehouse> elementInWarehouses,
                   List<ElementEvent> elementEvents, Attachment attachment, List<ElementsPlannedNumber> listOfElementsPlannedNumber) {
        this.id = id;
        this.name = name;
        this.code = code;
        this.typeOfUnit = typeOfUnit;
        this.quantityInUnit = quantityInUnit;
        this.elementReturnReleases = elementReturnReleases;
        this.elementInWarehouses = elementInWarehouses;
        this.elementEvents = elementEvents;
        this.attachment = attachment;
        this.listOfElementsPlannedNumber = listOfElementsPlannedNumber;
    }

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private String name;
    private String code;
    private TypeOfUnit typeOfUnit;
    private float quantityInUnit;
    private boolean deleted = Boolean.FALSE;

    @OneToMany(mappedBy = "element")
    private List<ElementReturnRelease> elementReturnReleases;

    @OneToMany(mappedBy = "element")
    private List<ElementInWarehouse> elementInWarehouses;

    @OneToMany(mappedBy = "element")
    private List<ElementEvent> elementEvents;

    @OneToOne
    private Attachment attachment;

    @OneToMany(mappedBy ="element" )
    private List<ElementsPlannedNumber> listOfElementsPlannedNumber;

    public int getInWarehouseCount(Long warehouseId) {
        Optional<ElementInWarehouse> elementInWarehouseOptional = elementInWarehouses.stream().filter(o -> o.getWarehouse().getId().equals(warehouseId)).findFirst();
        if(elementInWarehouseOptional.isPresent()) {
            return elementInWarehouseOptional.get().getInWarehouseCount();
        }else {
            throw new EntityNotFoundException();
        }
    }
}

File: Android/app/src/main/java/com/example/e_montazysta/data/model/Element.kt
package com.example.e_montazysta.data.model

import com.example.e_montazysta.data.repository.interfaces.IElementRepository
import com.example.e_montazysta.ui.element.ElementInWarehouse
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject

data class Element(
    val attachmentId: String?,
    val code: String,
    val deleted: Boolean,
    val elementEvents: List<Int>,
    val elementInWarehouses: List<ElementInWarehouse>?,
    val elementReturnReleases: List<Int>,
    val id: Int,
    val listOfElementsPlannedNumber: List<Int>,
    val name: String,
    val quantityInUnit: Double,
    val typeOfUnit: String
) {
    companion object : KoinComponent {
        suspend fun getElementDetails(elementId: Int): Element {
            val elementRepository: IElementRepository by inject()
            val result = elementRepository.getElementDetails(elementId)
            return when (result) {
                is Result.Success -> result.data
                is Result.Error -> throw result.exception
            }
        }
    }
}

